src/irr/asset/CGraphicsPipelineLoaderMTL.cpp

```diff
@@ -1,10 +1,393 @@
+    //extra info
+    uint extra;
+} PC;
+#endif //_IRR_FRAG_PUSH_CONSTANTS_DEFINED_
+
+#ifndef _IRR_FRAG_SET3_BINDINGS_DEFINED_
+#define _IRR_FRAG_SET3_BINDINGS_DEFINED_
+layout (set = 3, binding = 0) uniform sampler2D map_Ka;
+layout (set = 3, binding = 1) uniform sampler2D map_Kd;
+layout (set = 3, binding = 2) uniform sampler2D map_Ks;
+layout (set = 3, binding = 3) uniform sampler2D map_Ke;
+layout (set = 3, binding = 4) uniform sampler2D map_Ns;
+layout (set = 3, binding = 5) uniform sampler2D map_d;
+layout (set = 3, binding = 6) uniform sampler2D map_bump;
+#endif //_IRR_FRAG_SET3_BINDINGS_DEFINED_
+
+#include <irr/builtin/glsl/brdf/specular/fresnel/fresnel.glsl>
+
+#ifndef _IRR_FRAG_MAIN_DEFINED_
+#define _IRR_FRAG_MAIN_DEFINED_
+
+#define Ia 0.1
+void main()
+{
+    vec3 N = normalize(Normal);
+    if ((PC.extra&map_bump_MASK) == map_bump_MASK)
+    {
+        float height = texture(map_bump, UV).x;
+        vec3 dpdx = dFdx(ViewPos);
+        vec3 dpdy = dFdy(ViewPos);
+        float dhdx = dFdx(height);
+        float dhdy = dFdy(height);
+        
+        vec3 r1 = cross(dpdy, N);
+        vec3 r2 = cross(N, dpdx);
+        vec3 surfGrad = (r1*dhdx + r2*dhdy) / dot(dpdx,r1);
+        N = normalize(N - surfGrad);
+    }
+    vec3 L = normalize(-ViewPos);
+    vec3 R = normalize(-reflect(L,N));
+    float NdotL = max(dot(N,L), 0.0);
+    float VdotR = max(dot(L,R), 0.0);
+
+    vec3 Ke = texture(map_Ke, UV).rgb;//force visibility (as in: visible as resource used by the shader) of emissive map (needed temporarily because of bindings reordering bug)
+
+    vec3 Kd;
+    if ((PC.extra&(map_Kd_MASK)) == (map_Kd_MASK))
+        Kd = texture(map_Kd, UV).rgb;
+    else
+        Kd = PC.Kd;
+    float d = 1.0;
+    if ((PC.extra&(map_d_MASK)) == (map_d_MASK))
+        d = texture(map_d, UV).r;
+
+    vec3 color;
+    if ((PC.extra&ILLUM_MODEL_MASK) > 0)
+    {
+        vec3 Ka;
+        vec3 Ks;
+        float Ns;
+        if ((PC.extra&(map_Ka_MASK)) == (map_Ka_MASK))
+            Ka = texture(map_Ka, UV).rgb;
+        else
+            Ka = PC.Ka;
+        if ((PC.extra&(map_Ks_MASK)) == (map_Ks_MASK))
+            Ks = texture(map_Ks, UV).rgb;
+        else
+            Ks = PC.Ks;
+        if ((PC.extra&(map_Ns_MASK)) == (map_Ns_MASK))
+            Ns = texture(map_Ns, UV).x;
+        else
+            Ns = PC.Ns;
+
+        color = Ka*Ia + Kd*NdotL;
+        switch (PC.extra&ILLUM_MODEL_MASK)
+        {
+        case 2:
+        case 3://2 + reflection map
+        case 4://3 with transparency (glass)
+        case 6:
+        case 8://reflection map
+        case 9://reflection map
+            color += Ks*pow(VdotR, Ns);
+            break;
+        case 5:
+        case 7:
+            color += Ks*pow(VdotR, Ns)*Fresnel_dielectric(PC.Ni, NdotL);
+            break;
+        default:
+            break;
+        }
+    }
+    else color = Kd;
+
+    OutColor = vec4(color, d);
+}
+#endif //_IRR_FRAG_MAIN_DEFINED_
+)";
+}
+
```

Comment:
your fragment shader code is not dry

Reply:
why?

Reply:
hmmm I was thinking PBR and non-PBR were duplicated.

Reply:
glsl code seemed to repeat itself (might be due to presence of PBR shader)

Reply:
yes, the UV frag and non-UV frag (both non-PBR) BRDF code repeats itself.

Reply:
of course they are, they are different shaders

Reply:
you can make the non-UV version be an override of the UV version, you only need to redefine/override the main/lighting function and the FRAGMENT_INPUTS section + delete (just `#define _IRR_FRAG_SET3_BINDINGS_DEFINED_` without anything extra) the descriptor set 3 inputs

Reply:
thats exactly what it is for a few days now

Reply:
i dont know what repeats u're talking about then

Reply:
the BRDF code now.